src/mame/apple/tk10_keyboard.cpp

```diff
@@ -0,0 +1,233 @@
+// license:BSD-3-Clause
+// copyright-holders:Golden Child
+/**********************************************************************
+
+    TK-10 Apple ][ Clone keyboard
+
+**********************************************************************/
+
+#include "emu.h"
+
+#include "tk10_keyboard.h"
+
+#include "utf8.h"
+
+
+//**************************************************************************
+//  DEVICE DEFINITION
+//**************************************************************************
+
+// device type definition
+DEFINE_DEVICE_TYPE(TK10_KEYBOARD, tk10_keyboard_device, "tk10_keyboard", "TK-10 Keyboard")
+
+
+//-------------------------------------------------
+//  tk10_keyboard_device - constructor
+//-------------------------------------------------
+
+tk10_keyboard_device::tk10_keyboard_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, TK10_KEYBOARD, tag, owner, clock)
+	, m_data_write(*this)
+	, m_reset_write(*this)
+	, m_tk10cpu(*this, "tk10cpu")
+	, m_keyboard(*this, "P2%u", 0U)
+	, m_kbspecial(*this, "kbspecial")
+{
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+void tk10_keyboard_device::device_start()
+{
+}
+
+//-------------------------------------------------
+//  device_add_mconfig - device-specific config
+//-------------------------------------------------
+
+void tk10_keyboard_device::device_add_mconfig(machine_config &config)
+{
+	I8039(config, m_tk10cpu, 6_MHz_XTAL);
+	m_tk10cpu->set_addrmap(AS_PROGRAM, &tk10_keyboard_device::tk10_mem_map);
+
+	m_tk10cpu->t1_in_cb().set([this] () -> u8
+	{
+		u8 kbspecial = m_kbspecial->read();  // left and right shift keys
+		return (BIT(kbspecial, 0) & BIT(kbspecial, 1));
+	});
+
+	m_tk10cpu->t0_in_cb().set([this] () -> u8
+	{
+		return (BIT(m_kbspecial->read(), 2));  // control key
+	});
+
+	m_tk10cpu->p1_in_cb().set([this] () -> u8
+	{
+		u8 p2_r = m_tk10cpu->p2_r();
+		for (int i = 0; i < 8; i++)
+		{
+			if (!BIT(p2_r, i)) return m_keyboard[i]->read();
+			// read key matrix:
+			// a low value in p2 bit will drive bit of p1 low when switch contact made
+		}
```

Comment:
You need to AND together the active rows in a switch matrix.  It isn’t a multiplexer.

Reply:
I'm not quite sure that AND would work here in this situation.  

It treats P2 as a selector to select which column of the matrix is being scanned.

I got the schematic from:

http://john.ccac.rwth-aachen.de:8000/patrick/schematics.htm
http://john.ccac.rwth-aachen.de:8000/patrick/data/AppleClone_Keyboard.pdf

I did it that way because the code scans the p2 bits one by one, and then reads all of the p1 bits as a byte for the selected column.

I setup the PORT_BITS to match which bits of P1 will get driven low when P2 gets scanned.

 0:05A  mov  r6,#$3F                                        BE 3
 0:05C  mov  r2,#$7F                                        BA 7
 0:05E  mov  r4,#$08                                        BC 0
 0:060  mov  a,r2                                           FA
 0:061  rl   a                                              E7
 0:062  mov  r2,a                                           AA
 0:063  outl p2,a                                           3A
 0:064  in   a,p1                                           09
... loops back to 060

It writes a single bit low to p2, then reads p1.  The low bit in p2 rotates through (the RL A), doing the scanning .

Reply:
Yes, that’s a switch matrix.  When it’s trying to read one column at a time, it will drive a single column low, but think about what would happen if it drove multiple columns low at the same time.